lib/lightstep/tracer/tracer.rb

```diff
@@ -0,0 +1,374 @@
+    DEFAULT_MAX_REPORTING_PERIOD_SECS = 30.0
+
+    class Error < LightStep::Error; end
+    class ConfigurationError < LightStep::Tracer::Error; end
+
+    attr_reader :access_token, :guid
+
+    # Initialize a new tracer. Either an access_token or a transport must be
+    # provided. A component_name is always required.
+    # @param component_name [String] Component name to use for the tracer
+    # @param access_token [String] The project access token when pushing to LightStep
+    # @param transport [LightStep::Transport] How the data should be transported
+    # @return LightStep::Tracer
+    # @raise LightStep::ConfigurationError if the group name or access token is not a valid string.
+    def initialize(component_name:, access_token: nil, transport: nil)
+      configure(component_name: component_name, access_token: access_token, transport: transport)
+    end
+
+    def max_log_records
+      @max_log_records ||= DEFAULT_MAX_LOG_RECORDS
+    end
+
+    def max_log_records=(max)
+      @max_log_records = [MIN_MAX_LOG_RECORDS, max].max
+    end
+
+    def max_span_records
+      @max_span_records ||= DEFAULT_MAX_SPAN_RECORDS
+    end
+
+    def max_span_records=(max)
+      @max_span_records = [MIN_MAX_SPAN_RECORDS, max].max
+    end
+
+    def min_flush_period_micros
+      @min_flush_period_micros ||= DEFAULT_MIN_REPORTING_PERIOD_SECS * 1E6
+    end
+
+    def min_reporting_period_secs=(secs)
+      @min_flush_period_micros = [DEFAULT_MIN_REPORTING_PERIOD_SECS, secs].max * 1E6
+    end
+
+    def max_flush_period_micros
+      @max_flush_period_micros ||= DEFAULT_MAX_REPORTING_PERIOD_SECS * 1E6
+    end
+
+    def max_reporting_period_secs=(secs)
+      @max_flush_period_micros = [DEFAULT_MAX_REPORTING_PERIOD_SECS, secs].min * 1E6
+    end
+
+    # TODO(ngauthier@gmail.com) inherit SpanContext from references
+
+    # Starts a new span.
+    # @param operation_name [String] the operation name for the Span
+    # @param child_of [Span] Span to inherit from
+    # @param start_time [Time] When the Span started, if not now
+    # @param tags [Hash] tags for the span
+    # @return [Span]
+    def start_span(operation_name, child_of: nil, start_time: nil, tags: nil)
+      child_of_guid = nil
+      trace_guid = nil
+      if Span === child_of
+        child_of_guid = child_of.guid
+        trace_guid = child_of.trace_guid
+      else
+        trace_guid = generate_guid
+      end
+
+      Span.new(
+        tracer: self,
+        operation_name: operation_name,
+        child_of_guid: child_of_guid,
+        trace_guid: trace_guid,
+        start_micros: start_time.nil? ? now_micros : micros(start_time),
+        tags: tags
+      )
+    end
+
+    # Inject a span into the given carrier
+    # @param span [Span]
+    # @param format [LightStep::Tracer::FORMAT_TEXT_MAP, LightStep::Tracer::FORMAT_BINARY]
+    # @param carrier [Hash-like]
+    def inject(span, format, carrier)
+      case format
+      when LightStep::Tracer::FORMAT_TEXT_MAP
```

Comment:
we should support FORMAT_HTTP_HEADERS as well... for now it can share an impl with FORMAT_TEXT_MAP (which isn't totally correct but which will be fine modulo baggage, at least for the short term).